src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1662,6 +1662,82 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
             LowerFusedMultiplyAdd(node);
             break;
 
+        case NI_SSE_And:
+        case NI_SSE2_And:
+        case NI_AVX_And:
+        case NI_AVX2_And:
+        case NI_AVX512F_And:
+        case NI_AVX512DQ_And:
+        case NI_SSE_Or:
+        case NI_SSE2_Or:
+        case NI_AVX_Or:
+        case NI_AVX2_Or:
+        case NI_AVX512F_Or:
+        case NI_AVX512DQ_Or:
+        case NI_SSE_Xor:
+        case NI_SSE2_Xor:
+        case NI_AVX_Xor:
+        case NI_AVX2_Xor:
+        case NI_AVX512F_Xor:
+        case NI_AVX512DQ_Xor:
+        {
```

Comment:
Is `lowering` the right place for these rather than `morph`?

Reply:
There's some long term considerations around what it means to correctly support things like `kmask` and I expect its going to need several things that are currently delayed until lowering to be moved higher up. As we otherwise end up with missed opts due to things being put in locals or the like (which lowering can't handle).

We're ultimately going to need/want to handle things like ternary ops in VN as well as there are some core operations that would benefit from it.

Reply:
Thanks for the information! Do we expect another implementation of this optimization at `morph` as comparison?

Reply:
Hi @tannergooding, as I started to investigate this optimization during morph, a few things I'd like to share based on my internal tests:
1. VN will indeed fail to conduct constant folding in some cases, and hence lead to regression in terms of the code size.
2. I initially start with [fgOptimizeHWIntrinsic](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10654) at morph, one thing I am a bit concerning about is that at morph, we might miss some opt opportunities when 2 bitwise ops are in separate trees/statements (this is my naïve guess and I haven't verified if it is possible and how often this case might be hit.), while we might still be able to track them through UD-chain if implemented during lowering.
3. As some existing opts for vector bitwise AND in [fgOptimizeHWIntrinsic](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10799), we might be need to put 2 different possible opts under the same branch for AND, I wonder if it is expected, or if there is anywhere else we could put the opts.

Overall, I would be willing to put the code for implementation of this opts during `morph` into public for review if it is needed, and it will be appreciated if you could share more details on this, such that I can better understand the situation of this PR, thanks :)
> There's some long term considerations around what it means to correctly support things like `kmask` and I expect its going to need several things that are currently delayed until lowering to be moved higher up. As we otherwise end up with missed opts due to things being put in locals or the like (which lowering can't handle).

Reply:
> VN will indeed fail to conduct constant folding in some cases, and hence lead to regression in terms of the code size

I think this is something we want to add longer term. `TernaryLogic` is a pretty core API here.

> I initially start with [fgOptimizeHWIntrinsic](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10654) at morph, one thing I am a bit concerning about is that at morph, we might miss some opt opportunities when 2 bitwise ops are in separate trees/statements (this is my naïve guess and I haven't verified if it is possible and how often this case might be hit.), while we might still be able to track them through UD-chain if implemented during lowering.

Yes, there are likely some cases that can be missed here. The inverse is also true and CSE or other opts (locals being created) can make us miss this in lowering.

> I initially start with [fgOptimizeHWIntrinsic](https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10654) at morph, one thing I am a bit concerning about is that at morph, we might miss some opt opportunities when 2 bitwise ops are in separate trees/statements (this is my naïve guess and I haven't verified if it is possible and how often this case might be hit.), while we might still be able to track them through UD-chain if implemented during lowering.

Right. Its a general problem with recognizing different tree shapes and patterns that could occur.

> Overall, I would be willing to put the code for implementation of this opts during morph into public for review if it is needed, and it will be appreciated if you could share more details on this, such that I can better understand the situation of this PR, thanks :)

I don't think we need to block the current PR. Just raising that I think longer term and due to things like `KMask` and other support it may be desirable for it to exist in `morph` long term.

I believe that trying to propagate all the required information to lowering would be much more expensive/involved.